Introduce scheduled VM backups/snapshots

WebberHuang1118 opened the pull request:
**Description:**
Introduce scheduled VM backup, related HEP:https://github.com/harvester/harvester/pull/6008

**Solution:**
Implement schedule VM backup HEP

**Related Issue:**
#2756

**Test Plan: Backup Type**
- Prerequisites
  - Building Harvester Cluster with the master-head branch (`09f3ca` in my env)
  - Adding `ScheduleVMBackup` CRD
    - `$ kubectl apply -f deploy/charts/harvester-crd/templates/harvesterhci.io_schedulevmbackups.yaml` 
  - Updating `harvester` deployment from this PR
  - Updating `harvester-webhook` deployment from this PR
  - Configuring a valid S3 backup target
  - Creating and starting a VM `vm` in the default namespace

- Case0: Basic validation for CR `ScheduleVMBackup`
  - It will be rejected if the spec fields contain the following conditions
    - `spec.cron` with invalid cron format
    -  `spec.retain` < 2 or > 250
    - `spec.maxFailure` < 2 or >= `spec.retain` 
    - Backup target is not reachable (if the schedule is for backup type)

- Case1: Checking the backup rotation
  - Apply manifest to create a scheduled VM backup
    ```
    apiVersion: harvesterhci.io/v1beta1
    kind: ScheduleVMBackup
    metadata:
      annotations:
          harvesterhci.io/svmbackupSkipCronCheck: "enable"
      namespace: default
      name: svmbackup-vm
    spec:
      cron: '*/2 * * * *'
      retain: 3
      maxFailure: 2
      vmbackup:
        source:
          apiGroup: kubevirt.io
          kind: VirtualMachine
          name: vm
        type: backup
     ```
  - Checking new backups keep being created every two minutes
    - The Harvester always maintains at least three VM backups since the 6th minute
    - The oldest VM backup will be deleted as new VM backup completes
    - The oldest Longhorn snapshot should also be purged under the hood

- Case2: Scheduled backup for a stopped VM
  - Continuing the steps from Case1
  - Stopping VM `vm`
  - Checking new backups keep being created every two minutes
    - Related resources should keep being updated

- Case3: Scheduled backup keeps processing if temporal backup failure happens
  - Continuing the steps from Case1
  - Making the backup target unreachable
  - After a failure backup is discovered
    - `ScheduleVMBackup` CR `.status.failure = 1`
  - Making the backup target healthy
  - Checking new backups keep being created every two minutes
    - `ScheduleVMBackup` CR `.status.failure = 0`
    - Failure backup is removed
      - Not found in `ScheduleVMBackup` CR `.status.vmbackupInfo`

- Case4: #Failure backups reach threshold
  - Continuing the steps from Case3
  - Making the backup target unreachable
  - After around 8th minute
  - Schedules backup should be suspended
    -  `ScheduleVMBackup` CR `.status.failure = 2`
  - Checking new backups would not be created
    - `ScheduleVMBackup` CR `.status.suspend = true`
    - `ScheduleVMBackup` CR `.status.condition` contains the message `Reach Max Failure`
    - No more new VM backup is going to be created in this phase

- Case5: Resuming suspended backup schedule
  - Continuing the steps from Case4
  - Making the backup target healthy again
  - Setting `ScheduleVMBackup` CR `.spec.suspend = false`
  - Checking the backup schedule resume correctly
    - `ScheduleVMBackup` CR `.status.suspend = false`
    - `ScheduleVMBackup` CR `.status.failure = 0`
    - Checking the backup schedule resume correctly
      - New backups keep being created every two minutes

**Test Plan: Snapshot Type**
- Case1: Checking the snapshot rotation
  - Setting the backup target to default
  - Apply manifest to create a scheduled VM backup
    ```
    apiVersion: harvesterhci.io/v1beta1
    kind: ScheduleVMBackup
    metadata:
      annotations:
          harvesterhci.io/svmbackupSkipCronCheck: "enable"
      namespace: default
      name: svmbackup-snap
    spec:
      cron: '*/2 * * * *'
      retain: 3
      maxFailure: 2
      vmbackup:
        source:
          apiGroup: kubevirt.io
          kind: VirtualMachine
          name: vm
        type: snapshot
     ```
  - CR `ScheduleVMBackup` wouldn't be rejected even if the backup target is not reachable
  - Checking new snapshots keep being created every two minutes
    - The Harvester always maintains at least three VM snapshots since the 6th minute
    - The oldest VM snapshots will be deleted as new VM snapshot completes
    - The oldest Longhorn snapshots should also be purged under the hood

**Test Plan: Duplated  Schedule for the same VM**
- Creating two CR `ScheduleVMBackup` targets on the same VM
- The second one would be rejected by webhook
  - With the message `The request is invalid: spec.vmbackup: VM <VM> already has backup schedule` 

**Test Plan: Proactive Suspend/Resume the Schedule**
- Starting a scheduled VM backup or snapshot
- Editing the ScheduleVMBackup resource
  - As `spec.suspend` = true
  - No more new VM backups are created
- Editing the ScheduleVMBackup resource
  - As `spec.suspend` = false
  - New VM backup keeps being created in every two minutes

**Test Plan: Runtime update schedule cron expression**
- Creating a ScheduleVMBackup resource with backup or snapshot type
  - With `spec.vmbackup` points to the VM
  - With `spec.retain` = 3
  - With `spec.cron = */2 * * * *`
  - New VM backup keeps being created in every two minutes
- Editing the ScheduleVMBackup resource
  - As `spec.cron = */3 * * * *`
  - New VM backup keeps being created in every three minutes

**Test Plan: Schedule Granularity Check**
- Case1: If the schedule's granularity is less than one hour, it can't be applied
  - Apply manifest to create a scheduled VM backup
    ```
    apiVersion: harvesterhci.io/v1beta1
    kind: ScheduleVMBackup
    metadata:
      namespace: default
      name: svmbackup-vm1
    spec:
      cron: '*/2 * * * *'
      retain: 3
      maxFailure: 2
      vmbackup:
        source:
          apiGroup: kubevirt.io
          kind: VirtualMachine
          name: vm1
        type: backup
     ```
  - The creation would be ejected by webhook with `The request is invalid: spec.cron: schedule granularity 1m0s less than 1h0m0s`
- Case2: If the schedule's granularity is equal or more than one hour, it could be applied
  - Apply manifest to create a scheduled VM backup
    ```
    apiVersion: harvesterhci.io/v1beta1
    kind: ScheduleVMBackup
    metadata:
      namespace: default
      name: svmbackup-vm1
    spec:
      cron: '0 * * * *'
      retain: 3
      maxFailure: 2
      vmbackup:
        source:
          apiGroup: kubevirt.io
          kind: VirtualMachine
          name: vm1
        type: backup
     ```
  - The creation should be a success
- Case3: If two schedules have the same granularity and each iteration's timing offset is less than 10mins, the creation would be ejected
  - Following above plan
  - Apply manifest to create a new scheduled VM backup
    ```
    apiVersion: harvesterhci.io/v1beta1
    kind: ScheduleVMBackup
    metadata:
      namespace: default
      name: svmbackup-vm2
    spec:
      cron: '5 * * * *'
      retain: 3
      maxFailure: 2
      vmbackup:
        source:
          apiGroup: kubevirt.io
          kind: VirtualMachine
          name: vm2
        type: backup
     ```
  - The creation would be ejected by webhook with `The request is invalid: spec.cron: same granularity as schedule svmbackup-vm1 and offset less than 10m0s`
- Case4: If two schedules have the same granularity and each iteration's timing offset is *not* less than 10mins, the creation could be applied
  - Following above plan
  - Apply manifest to create a new scheduled VM backup
    ```
    apiVersion: harvesterhci.io/v1beta1
    kind: ScheduleVMBackup
    metadata:
      namespace: default
      name: svmbackup-vm2
    spec:
      cron: '10 * * * *'
      retain: 3
      maxFailure: 2
      vmbackup:
        source:
          apiGroup: kubevirt.io
          kind: VirtualMachine
          name: vm2
        type: backup
     ```
  - The creation should be a success
 
**Additional Context:**
- We can watch the VM backup list with:
  - `$ kubectl get schedulevmbackups.harvesterhci.io svmbackup-vm --watch -o json | jq -r .status`
- Since the GUI PR https://github.com/harvester/dashboard/pull/1103 is merged, the setting and getting operation for CR `ScheduleVMBackup` could be derived from GUI
- Annotation ` harvesterhci.io/svmbackupSkipCronCheck: "enable"` is used to skip granularity checking

FrankYang0529 reviewed (approved):
LGTM. Thanks for the PR.

WebberHuang1118 committed:
feat(schedule vm backup): Introduce scheduled VM backups

Improve webhook: exclusive schedule for VM and checking for VM snapshots
Support proactive suspend/resume, cron update and adjustment for LH v1.7.0
Implement Schedule Granularity Check

Signed-off-by: Webber Huang <webber.huang@suse.com>

tserong reviewed (approved):
LGTM in general, just a few small comments/suggestions/questions.

Also, the HEP doesn't seem to mention the granularity check, so that probably wants to be updated too.

bk201 reviewed (approved):
LGTM, thanks!
